### PR TITLE
Make generic client request function

### PR DESCRIPTION
**Pull request by cubisttriangle**
Also add some documentation and take client self as ref.

@elatif95, I tried to integrate a couple of the changes from your PR.  My hope there was to try to reduce conflicts either of us might have to deal with later.

@henryboisdequin  can you configure the repository so we can't merge if we're out of date with `main`? I think that may have caused some weirdness in the git history earlier.

Also, hi to both of you! Thanks for letting me join in!

### CODE REVIEW

File: src/url_builder.rs [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+            ("symbol", "GOOGL"),
+            ("from", "2020-12-10"),
+            ("to", "2021-01-10"),
+            ("token", token),
+        ];
+        let expected = format!("symbol=GOOGL&from=2020-12-10&to=2021-01-10&token={}", token);
+        let actual = UrlBuilder::join_params(&params);
+        assert_eq!(actual, expected);
+    }
+}
```
Comment (henryboisdequin) [inline src/url_builder.rs]: Small nit: Please add a new line at the end of the file

File: src/client.rs [hunk 2]
```diff
@@ -3,143 +3,136 @@
...
 use reqwest::Url;
+use crate::url_builder::UrlBuilder;
+use serde::de::DeserializeOwned;
 
 /// Finnhub API Client object.
 #[derive(Debug)]
 pub struct Client {
     /// API key from the Finnhub dashboard.
     pub api_key: String,
+    /// Constructs urls from root, endpoints, params
```
Comment (henryboisdequin) [inline src/client.rs]:
```suggestion
    /// Constructs urls from root, endpoints, params.
```

File: src/client.rs [hunk 3]
```diff
@@ -3,143 +3,136 @@
...
+    pub fn v1(api_key: String) -> Self {
+        Self {
+            api_key,
+            url_bldr: UrlBuilder::new("https://finnhub.io/api/v1"),
+        }
+    }
 
-        Ok(res)
+    /// Lookups a symbol in the Finnhub API
+    /// https://finnhub.io/docs/api/symbol-search
```
Comment (henryboisdequin) [inline src/client.rs:33]: Awesome!

**Review by henryboisdequin — APPROVED** [review]
Hi, your welcome! Thanks for this change, looks awesome. I love the idea of the URL builder, makes the code much cleaner and concise. I also like the docs of just adding the Finnhub documentation instead of summarising. 

I just have some small nits. Also, do you know how to configure the repo to make sure we can't merge if we are out of date with `main`? Correct me if I'm wrong but I thought this is the behaviour of GitHub by default.

File: src/client.rs (hunk 2 above)
Comment (cubisttriangle) [reply src/client.rs]: Added period.

File: src/url_builder.rs (hunk 1 above)
Comment (cubisttriangle) [reply src/url_builder.rs]: Added newline.

**Comment by cubisttriangle** [comment]
> Hi, your welcome! Thanks for this change, looks awesome. I love the idea of the URL builder, makes the code much cleaner and concise. I also like the docs of just adding the Finnhub documentation instead of summarising.
> 
> I just have some small nits. Also, do you know how to configure the repo to make sure we can't merge if we are out of date with `main`? Correct me if I'm wrong but I thought this is the behaviour of GitHub by default.

I am not too familiar with the actual configuration. It looks like you can add branch protection rules in the settings. You may be right about not allowing out of date commits to be merged. I just thought the commit history looked odd - where it shows my commit, then what appear to be duplicate commits for the merge.

**Comment by henryboisdequin** [comment]
> > Hi, your welcome! Thanks for this change, looks awesome. I love the idea of the URL builder, makes the code much cleaner and concise. I also like the docs of just adding the Finnhub documentation instead of summarising.
> > I just have some small nits. Also, do you know how to configure the repo to make sure we can't merge if we are out of date with `main`? Correct me if I'm wrong but I thought this is the behaviour of GitHub by default.
> 
> I am not too familiar with the actual configuration. It looks like you can add branch protection rules in the settings. You may be right about not allowing out of date commits to be merged. I just thought the commit history looked odd - where it shows my commit, then what appear to be duplicate commits for the merge.

Yes same, I'll look around. Once again, thanks for the PR :)

**Pull request merged**